src/probnum/randprocs/_random_process.py

```diff
@@ -0,0 +1,253 @@
+
+    def __repr__(self) -> str:
+        return (
+            f"<{self.__class__.__name__} with "
+            f"input_dim={self.input_dim}, output_dim={self.output_dim}, "
+            f"dtype={self.dtype}>"
+        )
+
+    @abc.abstractmethod
+    def __call__(self, x: _InputType) -> randvars.RandomVariable[_OutputType]:
+        """Evaluate the random process at a set of inputs.
+
+        Parameters
+        ----------
+        x
+            *shape=(input_dim,) or (n, input_dim)* -- Input(s) to evaluate random
+            process at.
+
+        Returns
+        -------
+        f
+            *shape=(), (output_dim,) or (n, output_dim)* -- Random process evaluated at
+            the inputs.
+        """
+        raise NotImplementedError
+
+    @property
+    def input_dim(self) -> int:
+        """Shape of inputs to the random process."""
+        return self._input_dim
+
+    @property
+    def output_dim(self) -> int:
+        """Shape of the random process evaluated at an input."""
+        return self._output_dim
+
+    @property
+    def dtype(self) -> np.dtype:
+        """Data type of (elements of) the random process evaluated at an input."""
+        return self._dtype
+
+    @abc.abstractmethod
+    def mean(self, x: _InputType) -> _OutputType:
+        """Mean function.
+
+        Returns the mean function evaluated at the given input(s).
+
+        Parameters
+        ----------
+        x
+            *shape=(input_dim,) or (n, input_dim)* -- Input(s) where the mean
+            function is evaluated.
+
+        Returns
+        -------
+        mean
+            *shape=(), (output_dim, ) or (n, output_dim)* -- Mean function of the
+            process evaluated at inputs ``x``.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def cov(self, x0: _InputType, x1: Optional[_InputType] = None) -> _OutputType:
+        r"""Covariance function or kernel.
+
+        Returns the covariance function :math:`\operatorname{Cov}(f(x_0),
+        f(x_1)) = \mathbb{E}[(f(x_0) - \mathbb{E}[f(x_0)])(f(x_0) - \mathbb{E}[f(
+        x_0)])^\top]` of the process evaluated at ``x0`` and ``x1``. If only ``x0`` is
+        given the covariance among the components of the random process at the
+        inputs defined by ``x0`` is computed.
+
+        Parameters
+        ----------
+        x0
+            *shape=(input_dim,) or (n0, input_dim)* -- First input to the covariance
+            function.
+        x1
+            *shape=(input_dim,) or (n1, input_dim)* -- Second input to the covariance
+            function.
+
+        Returns
+        -------
+        cov
+            *shape=(), (output_dim, output_dim), (n0, n1) or (n0, n1, output_dim,
+            output_dim)* -- Covariance of the process at ``x0`` and ``x1``. If
+            only ``x0`` is given the kernel matrix :math:`K=k(X_0, X_0)` is computed.
+        """  # pylint: disable=trailing-whitespace
+        raise NotImplementedError
+
```

Comment:
I think it would be quite useful to have a method along the lines of `marginal_cov(pointset)` which computes the covariances of all the respective points but not the covariance between points. This will be the biggest usecase for Filtering and ODESolver posteriors and having access to this method would allow implementing a fast way. 

Perhaps that is a pattern that we want to put more into the center? How about a method `self.marginal(x)` that computes the marginal distributions (i.e. kind of like a vectorised call, where the cross-terms are ignored). This will be 999/1000 usecases for GaussMarkov process calls, and since I agree that calling cov() on a set of points assembles the full, large covariance matrix, I'd like to have access to the other entries. In any case, if you agree that this would be fun, it suffices to place the interface here. I will take care of the GM implementation.

Reply:
(Oh and perhaps having a fast version of a `marginal`-like function could speed up `var` and `std` below.)

Reply:
That's a good idea. We could even extend this to `RandomVariables`. Calling `.marginal()` on a random variable returns a batch of random variables as discussed in the last software meeting.